File: app/src/main/java/com/star_zero/example/androidmvvm/presentation/tasks/TasksViewModel.java
package com.star_zero.example.androidmvvm.presentation.tasks;

import android.view.View;

import com.star_zero.example.androidmvvm.R;
import com.star_zero.example.androidmvvm.application.TaskService;
import com.star_zero.example.androidmvvm.domain.task.Task;
import com.star_zero.example.androidmvvm.presentation.tasks.adapter.TasksAdapter;

import javax.inject.Inject;

import rx.Observable;
import rx.subjects.PublishSubject;
import rx.subscriptions.CompositeSubscription;

public class TasksViewModel {

    private TaskService taskService;

    public final TasksAdapter adapter = new TasksAdapter();

    private CompositeSubscription subscriptions = new CompositeSubscription();

    @Inject
    TasksViewModel(TaskService taskService) {
        this.taskService = taskService;
    }

    void onCreate() {
        subscribe();
    }

    void onStart() {
        getTasks();
    }

    void onDestroy() {
        taskService.onDestroy();
        subscriptions.clear();
    }

    // ----------------------
    // Event
    // ----------------------

    public void onClickNewTask(View view) {
        clickNewTaskSubject.onNext(null);
    }

    // ----------------------
    // Operation
    // ----------------------

    private void getTasks() {
        taskService.getTasks();
    }

    void changeCompleteState(Task task, boolean completed) {
        taskService.changeCompleteState(task, completed);
    }

    // ----------------------
    // Subscribe
    // ----------------------

    private void subscribe() {
        subscriptions.add(taskService.tasks.subscribe(adapter::setTasks));
        subscriptions.add(taskService.errorGetTasks.subscribe(aVoid -> {
            errorMessageSubject.onNext(R.string.error_get_task);
        }));
        subscriptions.add(taskService.errorChangeCompleteState.subscribe(aVoid -> {
            errorMessageSubject.onNext(R.string.error_change_state);
        }));
    }

    // ----------------------
    // Notification
    // ----------------------

    private final PublishSubject<Void> clickNewTaskSubject = PublishSubject.create();
    final Observable<Void> clickNewTask = clickNewTaskSubject.asObservable();

    private final PublishSubject<Integer> errorMessageSubject = PublishSubject.create();
    final Observable<Integer> errorMessage = errorMessageSubject.asObservable();
}

File: app/src/main/kotlin/com/star_zero/example/androidmvvm/presentation/tasks/TasksViewModel.kt
package com.star_zero.example.androidmvvm.presentation.tasks

import android.arch.lifecycle.*
import android.view.View
import com.star_zero.example.androidmvvm.R
import com.star_zero.example.androidmvvm.application.TaskService
import com.star_zero.example.androidmvvm.domain.task.Task
import com.star_zero.example.androidmvvm.presentation.shared.viewmodel.ViewModelObservable
import com.star_zero.example.androidmvvm.utils.PublishLiveData
import com.star_zero.example.androidmvvm.utils.extension.fire
import com.star_zero.example.androidmvvm.utils.extension.observe
import javax.inject.Inject

class TasksViewModel @Inject constructor(val taskService: TaskService) : ViewModelObservable(), LifecycleObserver {

    private var owner: LifecycleOwner? = null

    // ----------------------
    // Lifecycle
    // ----------------------

    @OnLifecycleEvent(Lifecycle.Event.ON_CREATE)
    fun onCreate(owner: LifecycleOwner) {
        this.owner = owner
        subscribe(owner)
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_START)
    fun onStart() {
        fetchTasks()
    }

    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
    fun onDestroy() {
        owner = null
    }

    // ----------------------
    // Event
    // ----------------------

    fun onClickNewTask(@Suppress("UNUSED_PARAMETER") view: View) {
        clickNewTask.fire()
    }

    // ----------------------
    // Operation
    // ----------------------

    private fun fetchTasks() {
        taskService.fetchTasks(owner!!)
    }

    fun changeCompleteState(task: Task, completed: Boolean) {
        taskService.changeCompleteState(task, completed, owner!!)
    }

    // ----------------------
    // Subscribe
    // ----------------------

    private fun subscribe(owner: LifecycleOwner) {
        taskService.errorFetchTasks().observe(owner) {
            errorMessage.value = R.string.error_get_task
        }

        taskService.errorFetchTasks().observe(owner) {
            errorMessage.value = R.string.error_get_task
        }

        taskService.errorChangeCompleteState().observe(owner) {
            errorMessage.value = R.string.error_change_state
        }
    }

    // ----------------------
    // LiveData
    // ----------------------

    fun tasks(): LiveData<List<Task>> = Transformations.map(taskService.tasks(), { it })

    private val clickNewTask = PublishLiveData<Void>()
    fun clickNewTask(): LiveData<Void> = clickNewTask

    private val errorMessage = PublishLiveData<Int>()
    fun errorMessage(): LiveData<Int> = errorMessage
}

